### PR TITLE
FroSSL Implementation

### PR DESCRIPTION
**Pull request by OFSkean**
Implementation of FroSSL from https://arxiv.org/pdf/2310.02903

I am currently training a ResNet-18 on ImageNet100 to make sure everything is working properly

### CODE REVIEW

File: solo/losses/frossl.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    regularization_term = calculate_frobenius_regularization_term(z)
+    regularization_term = -1 * regularization_term # make sure its maximized
+
+    invariance_tradeoff = V * D * invariance_weight
+    invariance_term = calculate_invariance_term(z)
+    invariance_term = invariance_tradeoff * invariance_term
+
+    if logger is not None:
+        logger("frossl_regularization_loss", -regularization_term, sync_dist=True)
```
Comment (vturrisi) [inline solo/losses/frossl.py:85]: Is this supposed to be logged as `- regularization_term`? I would also say that for consistency with the other loss functions, it's better if you return the individual elements and log them on the method's side. You could return the two terms and sum them in the method.

**Comment by vturrisi** [comment]
@OFSkean Apart from my comment, the other parts look good. It's just missing the documentation for the method, results/checkpoints, and these to the README.

**Comment by jacobsn** [comment]
Hi @OFSkean — thanks for the original FroSSL implementation here. Since this PR has been inactive for a while, I've opened #398 which builds on your work and updates it to the current solo-learn API, addressing the review feedback from this thread:

- the loss function is now side-effect free and returns its individual terms (invariance + regularization), with logging done on the method side for consistency with the other losses;
- the optional `repitl` dependency is dropped (uses the paper's default linear kernel);
- README + docs entries and tests are added;
- reproduced CIFAR-10 (92.10%) / CIFAR-100 (69.52%) results are included.

Credit for the original implementation is yours. Happy to consolidate however the maintainers prefer.
